File: next-main/src/main/java/com/next/reservations/core/service/ReservationTimeConfigurationService.java
/*
package com.next.reservations.core.service;

import com.next.reservations.core.domain.Reservation;
import com.next.reservations.core.domain.ReservationTimeConfiguration;
import com.next.reservations.core.repository.ReservationTimeConfigurationRepository;
import org.springframework.stereotype.Service;

import javax.transaction.Transactional;
import java.sql.Date;
import java.time.LocalDate;
import java.util.Optional;

@Service
public class ReservationTimeConfigurationService {

    private ReservationTimeConfigurationRepository repository;

    public ReservationTimeConfigurationService(ReservationTimeConfigurationRepository repository) {
        this.repository = repository;
    }

    public ReservationTimeConfiguration createOrUpdate(Long id, String name, LocalDate defaultStartDate) {

        if(id == null){
            return create(name, defaultStartDate);
        } else {
            return update(id, name, defaultStartDate);
        }
    }

    private ReservationTimeConfiguration create(String name, LocalDate defaultStartDate) {
        final ReservationTimeConfiguration reservationTimeConfiguration =
                new ReservationTimeConfiguration(name, false, defaultStartDate);

        return repository.save(reservationTimeConfiguration);
    }

    private ReservationTimeConfiguration update(Long id, String name, LocalDate defaultStartDate) {
        final ReservationTimeConfiguration reservationTimeConfiguration = findById(id);

        reservationTimeConfiguration.setName(name);
        reservationTimeConfiguration.setDefaultStartDate(defaultStartDate);

        return repository.save(reservationTimeConfiguration);
    }

    @Transactional
    public void setAsDefault(Long id) {
        removePreviousDefault();

        final ReservationTimeConfiguration newDefault = findById(id);
        newDefault.setDefaultConfig(true);
        repository.save(newDefault);
    }


    private void removePreviousDefault() {
        final ReservationTimeConfiguration currentDefault = findByDefaultConfigTrue();

        currentDefault.setDefaultConfig(false);
        repository.save(currentDefault);
    }

    public ReservationTimeConfiguration findByDefaultConfigTrue() {
        return repository.findByDefaultConfigTrue();
    }

    public ReservationTimeConfiguration findById(Long id) {
        return repository.findById(id).orElseThrow();
    }

    public Optional<ReservationTimeConfiguration> findFutureDefault(){
        return repository.findByDefaultStartDateIsNotNull();
    }

}
*/

File: next-main/src/main/kotlin/com/next/reservations/core/service/ReservationTimeConfigurationService.kt
package com.next.reservations.core.service

import com.next.reservations.core.domain.ReservationTimeConfiguration
import com.next.reservations.core.repository.ReservationTimeConfigurationRepository
import org.springframework.stereotype.Service
import java.time.LocalDate
import javax.transaction.Transactional

@Service
class ReservationTimeConfigurationService(private val repository: ReservationTimeConfigurationRepository) {

    fun createOrUpdate(id: Long?, name: String, defaultStartDate: LocalDate?): ReservationTimeConfiguration {
        return id?.let { update(it, name, defaultStartDate) } ?: create(name, defaultStartDate)
    }

    private fun create(name: String, defaultStartDate: LocalDate?): ReservationTimeConfiguration {
        val reservationTimeConfiguration = ReservationTimeConfiguration(name, false, defaultStartDate)
        return repository.save(reservationTimeConfiguration)
    }

    private fun update(id: Long, name: String, defaultStartDate: LocalDate?): ReservationTimeConfiguration {
        val reservationTimeConfiguration = findById(id)
        reservationTimeConfiguration.name = name
        reservationTimeConfiguration.defaultStartDate = defaultStartDate
        return repository.save(reservationTimeConfiguration)
    }

    @Transactional
    fun setAsDefault(id: Long) {
        removePreviousDefault()
        val newDefault = findById(id)
        newDefault.defaultConfig = true
        newDefault.defaultStartDate = null
        repository.save(newDefault)
    }

    private fun removePreviousDefault() {
        val currentDefault = findByDefaultConfigTrue()
        currentDefault.defaultConfig = false
        repository.save(currentDefault)
    }

    fun findByDefaultConfigTrue(): ReservationTimeConfiguration {
        return repository.findByDefaultConfigTrue()
    }

    fun findById(id: Long): ReservationTimeConfiguration {
        return repository.findById(id).orElseThrow()
    }

    fun findFutureDefault(): ReservationTimeConfiguration? {
        return repository.findByDefaultStartDateIsNotNull()
    }

    fun removeFutureDefaultStartDate(){
        val futureDefault = findFutureDefault()
        if(futureDefault != null){
            futureDefault.defaultStartDate = null
            repository.save(futureDefault)
        }
    }

    fun setNewFutureDefaultStartDate(id: Long, newFutureDefaultStartDate: LocalDate){
        val config = findById(id)
        config.defaultStartDate = newFutureDefaultStartDate

        repository.save(config)
    }

    fun deleteById(id: Long) = repository.deleteById(id)

    fun findAll(): List<ReservationTimeConfiguration> =
            repository.findAll()

    fun validateCanDelete(id: Long) {
        val config = findById(id)
        if (config.defaultConfig)
            throw RuntimeException("You can't delete the default config!!!")
    }

    fun changeName(id: Long, name: String) {
        val config = findById(id)
        config.name = name
        repository.save(config)
    }

    fun addNew(name: String): ReservationTimeConfiguration {
        return repository.save(ReservationTimeConfiguration(
                name = name,
                defaultStartDate = null,
                defaultConfig = false
        ))
    }
}
